### PR TITLE
Create tinkerbell stack during cluster create

### PR DESCRIPTION
**Pull request by abhinavmpandey08**
Signed-off-by: Abhinav Pandey <abhinavmpandey08@gmail.com>

*Description of changes:*
This PR does the following:
- create the tinkerbell stack locally on the bootstrap cluster, with boots running on docker
- uses this temporary tinkerbell stack to provision a workload cluster
- moves the tinkerbell stack to this workload cluster

*Testing (if applicable):*
Ran create successfully and verified the workload cluster

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/providers/provider.go [hunk 1]
```diff
@@ -21,6 +21,7 @@ type Provider interface {
 	// PreCAPIInstallOnBootstrap is called after the bootstrap cluster is setup but before CAPI resources are installed on it. This allows us to do provider specific configuration on the bootstrap cluster.
 	PreCAPIInstallOnBootstrap(ctx context.Context, cluster *types.Cluster, clusterSpec *cluster.Spec) error
 	PostBootstrapSetup(ctx context.Context, clusterConfig *v1alpha1.Cluster, cluster *types.Cluster) error
+	PostWorkloadInit(ctx context.Context, cluster *types.Cluster, clusterSpec *cluster.Spec) error
```
Comment (chrisdoherty4) [inline pkg/providers/provider.go:24]: Needs a comment describing when it runs and what sort of activities should run (perhaps there are restrictions?)

File: pkg/providers/tinkerbell/tinkerbell.go [hunk 2]
```diff
@@ -316,7 +326,7 @@ func (p *Provider) InstallCustomProviderComponents(ctx context.Context, kubeconf
 	return nil
 }
 
-func (p *Provider) InstallTinkerbellStack(ctx context.Context, cluster *types.Cluster, clusterSpec *cluster.Spec) error {
+func (p *Provider) InstallTinkerbellStack(ctx context.Context, cluster *types.Cluster, clusterSpec *cluster.Spec, bootstrap bool) error {
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/tinkerbell.go:329]: Please can we rethink this. Passing bool flags to toggle behavior creates hard to follow branch logic that becomes increasingly complex with time.

File: pkg/providers/tinkerbell/tinkerbell.go [hunk 3]
```diff
@@ -71,7 +73,11 @@ type Provider struct {
 	Retrier          *retrier.Retrier
 }
 
-// TODO: Add necessary kubectl functions here
+type ProviderDockerClient interface {
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/tinkerbell.go:76]: I think this could be better named. Perhaps just `Docker` or `DockerRuntime`?

File: pkg/providers/tinkerbell/tinkerbell.go [hunk 4]
```diff
@@ -348,16 +358,43 @@ func (p *Provider) InstallTinkerbellStack(ctx context.Context, cluster *types.Cl
...
+		}
+
+		logger.V(5).Info("Running boots on docker for bootstrap")
+		flags := []string{
+			"-e", "DATA_MODEL_VERSION=kubernetes",
+			"-e", fmt.Sprintf("BOOTS_EXTRA_KERNEL_ARGS=tink_worker_image=%s", bundle.Tink.TinkWorker.URI),
+			// TODO: remove hardcoding when this PR is merged https://github.com/tinkerbell/boots/pull/273
+			"-e", "MIRROR_BASE_URL=https://tinkerbell-storage-for-eksa.s3.us-west-2.amazonaws.com",
+			"-e", fmt.Sprintf("TINKERBELL_GRPC_AUTHORITY=%s:42113", p.datacenterConfig.Spec.TinkerbellIP),
+			"-e", "TINKERBELL_TLS=false",
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/tinkerbell.go:391]: The hardcoded `-e` flags likely need to be configurable.

File: pkg/providers/tinkerbell/tinkerbell.go [hunk 5]
```diff
@@ -348,16 +358,43 @@ func (p *Provider) InstallTinkerbellStack(ctx context.Context, cluster *types.Cl
...
+			"-v", fmt.Sprintf("%s:/kubeconfig", kubeconfig),
+			"--network", "host",
+		}
+		cmd := []string{"-kubeconfig", "/kubeconfig", "-dhcp-addr", "0.0.0.0:67"}
+		if err := p.providerDockerClient.Run(ctx, bundle.Boots.Image.URI, bootsContainerName, cmd, flags...); err != nil {
+			return fmt.Errorf("running boots on docker: %v", err)
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/tinkerbell.go:397]:
```suggestion
			return fmt.Errorf("run boots with docker: %v", err)
```

File: pkg/providers/tinkerbell/tinkerbell.go [hunk 6]
```diff
@@ -329,16 +339,16 @@ func (p *Provider) InstallTinkerbellStack(ctx context.Context, cluster *types.Cl
...
-		// 	Name:        "boots",
-		// 	Manifest:    bundle.Boots.Manifest.URI,
-		// 	Deployments: []string{"boots"},
-		// },
-		// {
-		// 	Name:        "hegel",
-		// 	Manifest:    bundle.Hegel.Manifest.URI,
-		// 	Deployments: []string{"hegel"},
-		// },
+		{
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/tinkerbell.go:342]: I can't find something that actually uses this slice entry. The only place that seems to reference it is an explicit exclusion on l. 361.

File: pkg/providers/tinkerbell/create.go [hunk 7]
```diff
@@ -60,6 +60,22 @@ func (p *Provider) PostBootstrapSetup(ctx context.Context, clusterConfig *v1alph
 	return nil
 }
 
+func (p *Provider) PostWorkloadInit(ctx context.Context, cluster *types.Cluster, clusterSpec *cluster.Spec) error {
+	if p.setupTinkerbell {
+
+		logger.V(4).Info("Installing Tinkerbell stack on the workload cluster")
+		if err := p.InstallTinkerbellStack(ctx, cluster, clusterSpec, false); err != nil {
+			return fmt.Errorf("installing tinkerbell stack on the workload cluster: %v", err)
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/create.go:68]:
```suggestion
			return fmt.Errorf("install tinkerbell stack on workload cluster: %v", err)
```

File: pkg/providers/tinkerbell/create.go [hunk 8]
```diff
@@ -60,6 +60,22 @@ func (p *Provider) PostBootstrapSetup(ctx context.Context, clusterConfig *v1alph
...
+		}
+
+		logger.V(4).Info("Removing local boots container")
+		if err := p.providerDockerClient.ForceRemove(ctx, bootsContainerName); err != nil {
+			return fmt.Errorf("removing local boots container: %v", err)
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/create.go:73]:
```suggestion
			return fmt.Errorf("remove local boots container: %v", err)
```

**Review by chrisdoherty4 — CHANGES_REQUESTED** [review]

File: pkg/providers/tinkerbell/tinkerbell.go (hunk 3 above)
Comment (abhinavmpandey08) [reply pkg/providers/tinkerbell/tinkerbell.go:76]: I just followed how other executables are named like `ProviderKubectlClient`. I'm fine with just changing it to `Docker` but we should also change the others as well for consistency.

File: pkg/providers/tinkerbell/tinkerbell.go (hunk 6 above)
Comment (abhinavmpandey08) [reply pkg/providers/tinkerbell/tinkerbell.go:342]: The boots entry? It gets used when deploying boots on workload cluster

File: pkg/providers/tinkerbell/tinkerbell.go (hunk 2 above)
Comment (abhinavmpandey08) [reply pkg/providers/tinkerbell/tinkerbell.go:329]: Yeah, I am planning to rework this as part of my next PR which will change some of this logic. Are you okay with this for now?

File: pkg/providers/tinkerbell/tinkerbell.go [hunk 9]
```diff
@@ -325,18 +335,53 @@ func (p *Provider) InstallTinkerbellStack(ctx context.Context, cluster *types.Cl
 	}
 
 	for _, component := range components {
-		logger.V(5).Info("Applying manifest", "component", component.Name)
-		if err := p.providerKubectlClient.ApplyKubeSpec(ctx, cluster, component.Manifest); err != nil {
-			return fmt.Errorf("applying %s manifest: %v", component.Name, err)
-		}
+		if !bootstrap || component.Name != "boots" {
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/tinkerbell.go:338]:
```suggestion
		if !(bootstrap && component.Name != "boots") {
```
